rfc/rfc-89/rfc-89.md

```diff
@@ -0,0 +1,345 @@
+The read-write process of Spark based on Bucket Index is also similar.
+- Use `HoodieSimpleBucketIndex` to tag location.
+- Use `SparkBucketIndexPartitioner` to packs incoming records to be inserted into buckets (1 bucket = 1 RDD partition).
+- Use `BucketIndexSupport` to Bucket Index pruning during reading.
+
+## Design
+### Workflow
+The core process of writing to the Partition Level Bucket Index is shown in the figure below. First, the Bucket Identifier 
+loads the partition bucket count information from the metadata as a cache. During the Partitioner and Writer stages, when 
+using the Identifier for shuffling and tagging, it will first attempt to find the bucketNumber of the corresponding 
+partition from the cache. If the cache was missed, it will calculate the corresponding value based on the user's expression, 
+and then submit it to the metadata during the commit stage.
+![workflow.jpg](workflow.jpg)
+The design details of each part are as follows.
+
+### Config
+Add new Config `hoodie.bucket.index.partition.rule` default `regex`
+
+Add new config `hoodie.bucket.index.partition.expressions` default null
+
+Re-use existed config `hoodie.bucket.index.num.buckets` as default partition bucket number
+
+The above parameters are table-level parameters and need to be declared in the DDL. Users can't change this config through
+runtime options like Flink /*+ OPTIONS(...) */
+
+#### hoodie.bucket.index.partition.rule
+`hoodie.bucket.index.partition.rule` 
+
+It is used to set the rule parser. By default, it is a regular expression parser. 
+In the future, more parsers will be provided, such as range expression parsers and hybrid parsers. 
+Here, an interface-based capability will be offered, allowing users to customize parsers.
+
+#### hoodie.bucket.index.partition.expressions
+`hoodie.bucket.index.partition.expressions`
+
+Users can use this parameter to specify expression and the corresponding bucket numbers (separated by commas). 
+Multiple rules are separated by semicolons like `hoodie.bucket.index.partition.expressions=expression1,bucket-number1;expression2,bucket-number2`
+
+#### Demo
+```sql
+CREATE TABLE hudi_table(
+  id BIGINT,
+  name STRING,
+  price DOUBLE
+)
+WITH (
+'connector' = 'hudi',
+'path' = 'file:///tmp/hudi_table',
+'table.type' = 'MERGE_ON_READ',
+'index.type' = 'BUCKET',
+'hoodie.bucket.index.partition.expressions' = '\\d{4}-(06-(01|17|18)|11-(01|10|11)),256',
+'hoodie.bucket.index.num.buckets' = '10'
+);
+```
+
+For the dates of 06-01, 06-17, 06-18 in June and 01-11, 10-11, 11-11 in
+November of each year (in the format of yyyy-MM-dd), the corresponding bucket number for the partition is 256
+
+For common partitions use 10 as partition bucket number
+
+
+#### Config Management
+
+The expression config will be persisted as files named as `<instant>.hashing_config` for current table, it stores in
+`.hoodie/.simple_hashing/config` directory and contains the following information in json format
+
+```json
```

Comment:
Can you clarify why the expressions are stored in instant level, what is the cleaning strategy for legacy config files?